---
#### Author : Ariyan Nooshazar

#### Code Leader : Ariyan Nooshazar

#### Start date : 11/11/2022

#### Start date : 12/12/2022

#### Topic : League of Legends Winning Team Analysis & Game Predection

#### Email : ariyan.nooshazar@gmail.com

---

# Introduction




---
This notebook includes EDA and steps taken to clean our data and steps taken. We have collected our data from (1) - https://oracleselixir.com/tools/downloads, we have downloaded 3 datasets as CSV format for 3 consecutive years: 2020,2021 and 2022 for League of of Legends profesional games played.

In our project we are going to talk about, *League of Legends Winning Team Analysis & Game Prediction*. 

-- Questions of interest and full detailed report are included in the second notebook , alongside of an explanation of what is League of Legends.

---

| Syntax | Description |
| ----------- | ----------- |
| A | Total assists |
| AGT | Average game time/duration, in minutes |
| APG | Assists per game |
| B% | Percentage of games in which the champion was banned (not tied to a specific role) |
| BLND% | Blind-pick rate: percentage of games in which this player/champion was picked before their lane opponent (not always available) |
| BN% |Baron control rate |
| CCPM | Crowd control dealt to champions per minute |
| Champion | Champion name |
| CKPM | Average combined kills per minute (team kills + opponent kills) |
| CS%P15 | Average share of team's total CS post-15-minutes |
| CSD10 | Average creep score difference at 10 minutes |
| CSD15 | Average creep score difference at 15 minutes |
| CSD20 | Average creep score difference at 20 minutes |
| CSPM:| Average monsters + minions killed per minute |
| CTR% | Counter-pick rate: percentage of games in which this player/champion was picked after their lane opponent (not always available) |
| CWPM:| Control wards purchased per minute |
| D | Total deaths |
| D%P15 | Average share of team's damage to champions post-15-minutes |
| DMG% | Damage Share: average share of team’s total damage to champions |
| DMG%P15 | Average share of team's damage to champions post-15-minutes |
| DPG | Deaths per game |
| DPM | Average damage to champions per minute | 
| DRG% | Dragon control rate: percent of all Dragons killed that were taken by the team, reflecting only elemental drakes if ELD% is present | 
| DTH%	| Average share of team’s deaths | 
| EGPM	| Average earned gold per minute (excludes starting gold and inherent | gold generation)| 
| EGR | Early-Game Rating | 
| ELD% | Elder dragon control rate | 
| Event | Event name | 
| F3T% | First-to-three-towers rate (percentage of games in which team was the first to 3 tower kills| 
| FB% | First Blood rate -- for players/champions, percent of games earning a First Blood participation (kill or assist) |
| FBN%	| First Baron rate |
| FBV%	| First Blood Victim rate -- percent of games player/champion was killed for First Blood |
| FD%	| First dragon rate| 
| FT%	| First tower rate| 
| GD10	| Average gold difference at 10 minutes | 
| GD15	| Average gold difference at 15 minutes | 
| GD20	| Average gold difference at 20 minutes | 
| GOLD%	| Gold Share: average share of team’s total gold earned (excludes starting gold and inherent gold generation) | 
| GP	| Games Played | 
| GPM	| Average gold per minute | 
| GPR	| Gold percent rating (average amount of game’s total gold held, relative to 50%) | 
| GSPD	| Average gold spent percentage difference| 
| GXD10	| Average gold+experience difference at 10 minutes | 
| GXD15	| Average gold+experience difference at 15 minutes | 
| GXD20	| Average gold+experience difference at 20 minutes | 
| HLD%	| Rift Herald control rate| 
| IWC%	| Average percentage of opponent’s invisible wards cleared | 
| JNG%	| Jungle Control: average share of game’s total jungle CS | 
| K	| Total kills |  
| KD | Kill-to-Death Ratio | 
| KDA | Total Kill/Death/Assist ratio |
| KP | Kill participation: percentage of team's kills in which player earned a Kill or Assist |
| KPG | Kills per game |
| KS% | Kill share: player's percentage of their team's total kills |
| L | Losses |
| LNE% | Lane Control: average share of game’s total lane CS |
| Losse | 	Total Losses |
| LP | Ladder Points |
| MLR | Mid/Late Rating |
| OE | Rating	Oracle’s Elixir Performance Rating |
| O |  Rtg	Oracle’s Elixir Performance Rating |
| P% | Percentage of games champion was picked in this role. |
| P+B% | Percentage of games in which the champion was either banned or picked in any role |
| Playe | Player's in-game name |
| Pos | Position |
| PPG | Turret plates destroyed per game |
| Rank | Official Leaderboard Rank |
| STL | Neutral objectives stolen |
| STLP | Neutral objectives stolen per game |
| StPG | Neutral objectives stolen per game |
| Team | Team name |
| VSPM | Vision score per minute |
| VWC% | Average percentage of opponent’s visible wards cleared |
| W	| Wins |
| W% | Win percentage |
| WC% | Average percentage of opponent wards cleared |
| WCPM | Average wards cleared per minute |
| Wins | Total Wins |
| WPM | Average wards placed per minute |
| XPD10 | Average experience difference at 10 minutes |
| XPD15 | Average experience difference at 15 minutes |
| XPD20 | Average experience difference at 20 minutes |

In [ ]:
# Basic packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os  

---
### Load The Data

In [ ]:
# loading our 3 data sets
df_20 = pd.read_csv("data/2020_LoL_esports_match_data_from_OraclesElixir_20221125.csv")
df_21 = pd.read_csv("data/2021_LoL_esports_match_data_from_OraclesElixir_20221125.csv")
df_22 = pd.read_csv("data/2022_LoL_esports_match_data_from_OraclesElixir_20221125.csv")


C:\Users\Ariyan\AppData\Local\Temp\ipykernel_12284\662293833.py:4: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  df_22 = pd.read_csv("data/2022_LoL_esports_match_data_from_OraclesElixir_20221125.csv")


In [ ]:
# display our dataset
df_20.head()

gameid datacompleteness  \
0  ESPORTSTMNT03/1241318         complete   
1  ESPORTSTMNT03/1241318         complete   
2  ESPORTSTMNT03/1241318         complete   
3  ESPORTSTMNT03/1241318         complete   
4  ESPORTSTMNT03/1241318         complete   

                                                 url league  year split  \
0  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   NaN   
1  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   NaN   
2  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   NaN   
3  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   NaN   
4  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   NaN   

   playoffs                 date  game  patch  ...  opp_csat15 golddiffat15  \
0         0  2020-01-03 07:33:26     1   9.24  ...       118.0        165.0   
1         0  2020-01-03 07:33:26     1   9.24  ...        98.0       -399.0   
2         0  2020-01-03 07:33:26     1   9.24  ...       140.0       -409.0   
3         0  2020-01-03 07:33:26     1   9.24  ...       135.0         51.0   
4         0  2020-01-03 07:33:26     1   9.24  ...        28.0       -233.0   

  xpdiffat15 csdiffat15 killsat15 assistsat15 deathsat15 opp_killsat15  \
0      166.0       13.0       0.0         0.0        0.0           0.0   
1      150.0       -7.0       0.0         0.0        0.0           1.0   
2    -1837.0      -11.0       0.0         0.0        1.0           0.0   
3     -401.0        1.0       0.0         0.0        0.0           0.0   
4      257.0       -8.0       0.0         0.0        0.0           0.0   

  opp_assistsat15 opp_deathsat15  
0             0.0            0.0  
1             0.0            0.0  
2             1.0            0.0  
3             0.0            0.0  
4             1.0            0.0  

[5 rows x 123 columns]

In [ ]:
# display our dataset
df_21.head()

gameid datacompleteness  \
0  ESPORTSTMNT03/1632489         complete   
1  ESPORTSTMNT03/1632489         complete   
2  ESPORTSTMNT03/1632489         complete   
3  ESPORTSTMNT03/1632489         complete   
4  ESPORTSTMNT03/1632489         complete   

                                                 url league  year split  \
0  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   
1  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   
2  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   
3  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   
4  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   

   playoffs                 date  game  patch  ...  opp_csat15 golddiffat15  \
0         0  2021-01-02 07:40:39   1.0  10.25  ...       118.0        748.0   
1         0  2021-01-02 07:40:39   1.0  10.25  ...        84.0       2120.0   
2         0  2021-01-02 07:40:39   1.0  10.25  ...       143.0       1578.0   
3         0  2021-01-02 07:40:39   1.0  10.25  ...       120.0        124.0   
4         0  2021-01-02 07:40:39   1.0  10.25  ...        25.0        448.0   

  xpdiffat15 csdiffat15 killsat15 assistsat15 deathsat15 opp_killsat15  \
0      -56.0       -4.0       2.0         0.0        1.0           0.0   
1     3405.0       62.0       3.0         2.0        0.0           2.0   
2      354.0       15.0       2.0         3.0        0.0           0.0   
3      102.0       10.0       0.0         4.0        2.0           1.0   
4      450.0        3.0       1.0         4.0        0.0           0.0   

  opp_assistsat15 opp_deathsat15  
0             1.0            1.0  
1             0.0            3.0  
2             0.0            0.0  
3             1.0            2.0  
4             2.0            2.0  

[5 rows x 123 columns]

In [ ]:
# display our dataset
df_22.head()

gameid datacompleteness  url  league  year   split  \
0  ESPORTSTMNT01_2690210         complete  NaN  LCK CL  2022  Spring   
1  ESPORTSTMNT01_2690210         complete  NaN  LCK CL  2022  Spring   
2  ESPORTSTMNT01_2690210         complete  NaN  LCK CL  2022  Spring   
3  ESPORTSTMNT01_2690210         complete  NaN  LCK CL  2022  Spring   
4  ESPORTSTMNT01_2690210         complete  NaN  LCK CL  2022  Spring   

   playoffs                 date  game  patch  ...  opp_csat15 golddiffat15  \
0         0  2022-01-10 07:44:08     1  12.01  ...       121.0        391.0   
1         0  2022-01-10 07:44:08     1  12.01  ...       100.0        541.0   
2         0  2022-01-10 07:44:08     1  12.01  ...       119.0       -475.0   
3         0  2022-01-10 07:44:08     1  12.01  ...       149.0       -793.0   
4         0  2022-01-10 07:44:08     1  12.01  ...        21.0        443.0   

  xpdiffat15 csdiffat15 killsat15 assistsat15 deathsat15 opp_killsat15  \
0      345.0       14.0       0.0         1.0        0.0           0.0   
1     -275.0      -11.0       2.0         3.0        2.0           0.0   
2      153.0        1.0       0.0         3.0        0.0           3.0   
3    -1343.0      -34.0       2.0         1.0        2.0           3.0   
4     -497.0        7.0       1.0         2.0        2.0           0.0   

  opp_assistsat15 opp_deathsat15  
0             1.0            0.0  
1             5.0            1.0  
2             3.0            2.0  
3             3.0            0.0  
4             6.0            2.0  

[5 rows x 123 columns]

In [ ]:
# Shape of our data
df_20.shape

(118668, 123)

In [ ]:
# Shape of our data
df_21.shape

(150264, 123)

In [ ]:
# Shape of our data
df_22.shape

(147984, 123)

In [ ]:
# Checking for duplicates
df_20.duplicated().sum()

0

In [ ]:
# Checking for duplicates
df_21.duplicated().sum()

0

In [ ]:
# Checking for duplicates
df_22.duplicated().sum()

0

Seems like we have the luxury to work with a semi-clean data, that will make our job much easier. Above we checked for duplicates and there were no duplicates in any of our dataset.

We also checked for shape of all 3 datasets, and all 3 data sets have the exact same amount of features and are identical, with each one having over 100,000 rows, being a big data size.

Now that we know there are no duplicates and all features of our datasets are the same, we will proceed to  merge our datasets to one;

---
### Merging Our Datas

In [ ]:
# Merging our Datas
M1_df = pd.concat([df_20, df_21, df_22], axis=0)
M1_df

gameid datacompleteness  \
0       ESPORTSTMNT03/1241318         complete   
1       ESPORTSTMNT03/1241318         complete   
2       ESPORTSTMNT03/1241318         complete   
3       ESPORTSTMNT03/1241318         complete   
4       ESPORTSTMNT03/1241318         complete   
...                       ...              ...   
147979  ESPORTSTMNT03_3062711         complete   
147980  ESPORTSTMNT03_3062711         complete   
147981  ESPORTSTMNT03_3062711         complete   
147982  ESPORTSTMNT03_3062711         complete   
147983  ESPORTSTMNT03_3062711         complete   

                                                      url league  year  \
0       http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   
1       http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   
2       http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   
3       http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   
4       http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2020   
...                                                   ...    ...   ...   
147979                                                NaN   NEXO  2023   
147980                                                NaN   NEXO  2023   
147981                                                NaN   NEXO  2023   
147982                                                NaN   NEXO  2023   
147983                                                NaN   NEXO  2023   

          split  playoffs                 date  game  patch  ...  opp_csat15  \
0           NaN         0  2020-01-03 07:33:26   1.0   9.24  ...       118.0   
1           NaN         0  2020-01-03 07:33:26   1.0   9.24  ...        98.0   
2           NaN         0  2020-01-03 07:33:26   1.0   9.24  ...       140.0   
3           NaN         0  2020-01-03 07:33:26   1.0   9.24  ...       135.0   
4           NaN         0  2020-01-03 07:33:26   1.0   9.24  ...        28.0   
...         ...       ...                  ...   ...    ...  ...         ...   
147979  Split 1         1  2022-11-22 16:35:58   3.0  12.21  ...       146.0   
147980  Split 1         1  2022-11-22 16:35:58   3.0  12.21  ...       118.0   
147981  Split 1         1  2022-11-22 16:35:58   3.0  12.21  ...        23.0   
147982  Split 1         1  2022-11-22 16:35:58   3.0  12.21  ...       531.0   
147983  Split 1         1  2022-11-22 16:35:58   3.0  12.21  ...       492.0   

       golddiffat15 xpdiffat15 csdiffat15 killsat15 assistsat15 deathsat15  \
0             165.0      166.0       13.0       0.0         0.0        0.0   
1            -399.0      150.0       -7.0       0.0         0.0        0.0   
2            -409.0    -1837.0      -11.0       0.0         0.0        1.0   
3              51.0     -401.0        1.0       0.0         0.0        0.0   
4            -233.0      257.0       -8.0       0.0         0.0        0.0   
...             ...        ...        ...       ...         ...        ...   
147979        199.0       87.0        7.0       2.0         4.0        0.0   
147980        675.0      -22.0       -1.0       2.0         2.0        2.0   
147981         72.0      432.0        0.0       0.0         6.0        1.0   
147982      -2693.0    -2278.0      -39.0       4.0         9.0        7.0   
147983       2693.0     2278.0       39.0       7.0        16.0        4.0   

       opp_killsat15 opp_assistsat15 opp_deathsat15  
0                0.0             0.0            0.0  
1                1.0             0.0            0.0  
2                0.0             1.0            0.0  
3                0.0             0.0            0.0  
4                0.0             1.0            0.0  
...              ...             ...            ...  
147979           2.0             0.0            2.0  
147980           1.0             3.0            2.0  
147981           0.0             4.0            2.0  
147982           7.0            16.0            4.0  
147983           4.0             9.0     

In [ ]:
# Checking shape of our merged data
M1_df.shape

(416916, 123)

After merging our datasets into one dataframe we end up with 416,916 rows and a total of 123 features.

In our data we came across some rows that have partial informations and rows that have name of players of each team; we do not need partial information nor information of players as we trying to build model based on the result of the game and in such case, the rows that contain the name of team is be enough.

---

In [ ]:
# dropping partial rows
index_names = M1_df[M1_df['datacompleteness'] == 'partial'].index

In [ ]:
M1_df.drop(index_names, inplace = True)

In [ ]:
# dropping rows that do not contain any info about team
index_names_1 = M1_df[M1_df['position'] != 'team'].index

In [ ]:
M1_df.drop(index_names_1, inplace = True)

In [ ]:
# Sanity Check
M1_df.shape

(54258, 123)

---
**As seen below, we are listing all the columns we have and determining which features we do not need based on the models we are going to make.**

---

In [ ]:
# Listing name of our features
M1_df.columns[M1_df.any()].tolist()

['gameid',
 'datacompleteness',
 'url',
 'league',
 'year',
 'split',
 'playoffs',
 'date',
 'game',
 'patch',
 'participantid',
 'side',
 'position',
 'teamname',
 'teamid',
 'ban1',
 'ban2',
 'ban3',
 'ban4',
 'ban5',
 'gamelength',
 'result',
 'kills',
 'deaths',
 'assists',
 'teamkills',
 'teamdeaths',
 'doublekills',
 'triplekills',
 'quadrakills',
 'pentakills',
 'firstblood',
 'team kpm',
 'ckpm',
 'firstdragon',
 'dragons',
 'opp_dragons',
 'elementaldrakes',
 'opp_elementaldrakes',
 'infernals',
 'mountains',
 'clouds',
 'oceans',
 'chemtechs',
 'hextechs',
 'dragons (type unknown)',
 'elders',
 'opp_elders',
 'firstherald',
 'heralds',
 'opp_heralds',
 'firstbaron',
 'barons',
 'opp_barons',
 'firsttower',
 'towers',
 'opp_towers',
 'firstmidtower',
 'firsttothreetowers',
 'turretplates',
 'opp_turretplates',
 'inhibitors',
 'opp_inhibitors',
 'damagetochampions',
 'dpm',
 'damagetakenperminute',
 'damagemitigatedperminute',
 'wardsplaced',
 'wpm',
 'wardskilled',
 'wcpm',
 '

In [ ]:
# dropping the columns we do not need
M1_df.drop(['total cs','earnedgoldshare','damageshare','deathsat15','deathsat10','teamdeaths','url','date','gamelength','doublekills','split', 'champion','gameid','game', 'patch','triplekills','quadrakills','pentakills', 'participantid','playername', 'playerid', 'split','position', 'teamid', 'champion','ban1','ban2', 'ban3', 'ban4', 'ban5','opp_dragons','opp_elementaldrakes', 'opp_elders','opp_heralds', 'opp_barons', 'opp_towers', 'opp_turretplates', 'opp_inhibitors', 'opp_goldat10','opp_xpat10','opp_csat10', 'opp_killsat10','opp_assistsat10','opp_deathsat10', 'opp_killsat15','opp_goldat15','opp_xpat15','opp_csat15','opp_assistsat15','opp_deathsat15','ckpm','monsterkillsenemyjungle','monsterkillsownjungle','datacompleteness','league','playoffs','dpm','damagetochampions','gspd','vspm','visionscore','wpm','wardskilled','wcpm','wardsplaced','damagemitigatedperminute','damagetakenperminute','deaths','year','firstbloodassist','firstbloodvictim','firstbloodkill',  'team kpm', 'elementaldrakes',  ], axis=1, inplace = True)

In [ ]:
# Sanity Check
M1_df.shape

(54258, 51)

Please note once we merged and dropped the names of the players per team, we noticed that the size of our data significantly dropped to 54,258 rows from 416,916 rows; however, we are not worried as we have fully detailed information of ~55,000 games in our dataset.

We are analyzing the game based on two topics;
1. Gold earned on the first 15 minutes of a match 


2. Secondary objectives in the game(turrets, dragons and etc...)

Therefore we dropped columns that were not answering or providing us any information about the questions above, or that could possibly dominate other columns once we start building our models.

In [ ]:
#  listing columns we have
M1_df.columns[M1_df.any()].tolist()

['side',
 'teamname',
 'result',
 'kills',
 'assists',
 'teamkills',
 'firstblood',
 'firstdragon',
 'dragons',
 'infernals',
 'mountains',
 'clouds',
 'oceans',
 'chemtechs',
 'hextechs',
 'dragons (type unknown)',
 'elders',
 'firstherald',
 'heralds',
 'firstbaron',
 'barons',
 'firsttower',
 'towers',
 'firstmidtower',
 'firsttothreetowers',
 'turretplates',
 'inhibitors',
 'controlwardsbought',
 'totalgold',
 'earnedgold',
 'earned gpm',
 'goldspent',
 'minionkills',
 'monsterkills',
 'cspm',
 'goldat10',
 'xpat10',
 'csat10',
 'golddiffat10',
 'xpdiffat10',
 'csdiffat10',
 'killsat10',
 'assistsat10',
 'goldat15',
 'xpat15',
 'csat15',
 'golddiffat15',
 'xpdiffat15',
 'csdiffat15',
 'killsat15',
 'assistsat15']

In [ ]:
# display our data
M1_df.head()

side      teamname  result  kills  assists  teamkills  firstblood  \
10  Blue  Liiv SANDBOX       1     17       33         17         0.0   
11   Red            T1       0     11       25         11         1.0   
22  Blue            T1       0      5       16          5         0.0   
23   Red  Liiv SANDBOX       1     21       41         21         1.0   
46  Blue            T1       0      4        8          4         0.0   

    firstdragon  dragons  infernals  ...  killsat10  assistsat10  goldat15  \
10          0.0      2.0        0.0  ...        0.0          0.0   21953.0   
11          1.0      3.0        1.0  ...        0.0          0.0   22778.0   
22          0.0      1.0        0.0  ...        0.0          0.0   21714.0   
23          1.0      4.0        0.0  ...        3.0          3.0   27198.0   
46          1.0      2.0        0.0  ...        0.0          0.0   22329.0   

     xpat15  csat15  golddiffat15  xpdiffat15  csdiffat15  killsat15  \
10  27219.0   507.0        -825.0     -1665.0       -12.0        0.0   
11  28884.0   519.0         825.0      1665.0        12.0        1.0   
22  26869.0   501.0       -5484.0     -4833.0       -56.0        0.0   
23  31702.0   557.0        5484.0      4833.0        56.0        6.0   
46  26093.0   465.0       -2471.0     -3064.0       -70.0        2.0   

    assistsat15  
10          0.0  
11          2.0  
22          0.0  
23          7.0  
46          3.0  

[5 rows x 51 columns]

We have now merged our data and dropped all the features we didn't need; therefore, let's start the EDA on our merged dataframe by checking if there is any null values.

In [ ]:
# Checking for null values
M1_df.isna().sum()

side                          0
teamname                     27
result                        0
kills                         0
assists                       0
teamkills                     0
firstblood                   22
firstdragon                   6
dragons                      68
infernals                    68
mountains                    68
clouds                       68
oceans                       68
chemtechs                 32202
hextechs                  32202
dragons (type unknown)    51115
elders                       68
firstherald                   0
heralds                    2534
firstbaron                  354
barons                        0
firsttower                   22
towers                        0
firstmidtower                22
firsttothreetowers            0
turretplates              32202
inhibitors                   76
controlwardsbought           22
totalgold                     0
earnedgold                   22
earned gpm                   22
goldspen

---
There are some columns that appear as if they were missing a lot of value like `hexctechs`(type of dragon) or `heralds`, well the truth is they are not really missing any value, simply put: the value for those columns are 0, which means they have not obtained any of those resources in the game. For these type of columns we have decided to place the null value with 0.

Also there are some other columns like `teamname` or `xpat15` that are columns similar to them with such a low variables missing like 2 rows or 22, but provided that it is basically impossible to have no experience(xp15) at 15 minutes, and they are missing such small amount of value, and in that case their relevance in our analysis would be minimal so we have decided to drop them .

---

In [ ]:
# Dropping null value rows
M1_df.dropna( how='any',
                          subset=['teamname', 'golddiffat15', 'assistsat15', 'csdiffat15', 'killsat15', 'xpdiffat15','csat15','xpat15','goldat15','earnedgold','earned gpm','firstblood'], inplace = True)

In [ ]:
# Sanity Check
M1_df.isna().sum()

side                          0
teamname                      0
result                        0
kills                         0
assists                       0
teamkills                     0
firstblood                    0
firstdragon                   6
dragons                      68
infernals                    68
mountains                    68
clouds                       68
oceans                       68
chemtechs                 32180
hextechs                  32180
dragons (type unknown)    51086
elders                       68
firstherald                   0
heralds                    2512
firstbaron                  352
barons                        0
firsttower                    0
towers                        0
firstmidtower                20
firsttothreetowers            0
turretplates              32180
inhibitors                   76
controlwardsbought            0
totalgold                     0
earnedgold                    0
earned gpm                    0
goldspen

In [ ]:
# filling null values with 0
M1_df=M1_df.fillna(0)

In [ ]:
# Sanity Check
M1_df.isna().sum()

side                      0
teamname                  0
result                    0
kills                     0
assists                   0
teamkills                 0
firstblood                0
firstdragon               0
dragons                   0
infernals                 0
mountains                 0
clouds                    0
oceans                    0
chemtechs                 0
hextechs                  0
dragons (type unknown)    0
elders                    0
firstherald               0
heralds                   0
firstbaron                0
barons                    0
firsttower                0
towers                    0
firstmidtower             0
firsttothreetowers        0
turretplates              0
inhibitors                0
controlwardsbought        0
totalgold                 0
earnedgold                0
earned gpm                0
goldspent                 0
minionkills               0
monsterkills              0
cspm                      0
goldat10            

---
We had to rest our indexes as it created extra column final look of our cleaned merged dta frame is as it looks below;

In [ ]:
# reset our index 
M1_df = M1_df.sort_values(by=[]).reset_index().drop('index', axis=1)
M1_df

side       teamname  result  kills  assists  teamkills  firstblood  \
0      Blue   Liiv SANDBOX       1     17       33         17         0.0   
1       Red             T1       0     11       25         11         1.0   
2      Blue             T1       0      5       16          5         0.0   
3       Red   Liiv SANDBOX       1     21       41         21         1.0   
4      Blue             T1       0      4        8          4         0.0   
...     ...            ...     ...    ...      ...        ...         ...   
54202   Red    QLASH Spain       1     22       26         22         1.0   
54203  Blue   Kawaii Kiwis       1     28       70         28         1.0   
54204   Red  Stormbringers       0     10       24         10         0.0   
54205  Blue           ZEST       1     23       70         23         0.0   
54206   Red    QLASH Spain       0     13       35         13         1.0   

       firstdragon  dragons  infernals  ...  killsat10  assistsat10  goldat15  \
0              0.0      2.0        0.0  ...        0.0          0.0   21953.0   
1              1.0      3.0        1.0  ...        0.0          0.0   22778.0   
2              0.0      1.0        0.0  ...        0.0          0.0   21714.0   
3              1.0      4.0        0.0  ...        3.0          3.0   27198.0   
4              1.0      2.0        0.0  ...        0.0          0.0   22329.0   
...            ...      ...        ...  ...        ...          ...       ...   
54202          1.0      4.0        0.0  ...        3.0          2.0   24907.0   
54203          0.0      2.0        0.0  ...        7.0         12.0   26400.0   
54204          1.0      3.0        1.0  ...        2.0          3.0   24672.0   
54205          0.0      2.0        0.0  ...        2.0          4.0   23741.0   
54206          1.0      3.0        0.0  ...        3.0          6.0   26434.0   

        xpat15  csat15  golddiffat15  xpdiffat15  csdiffat15  killsat15  \
0      27219.0   507.0        -825.0     -1665.0       -12.0        0.0   
1      28884.0   519.0         825.0      1665.0        12.0        1.0   
2      26869.0   501.0       -5484.0     -4833.0       -56.0        0.0   
3      31702.0   557.0        5484.0      4833.0        56.0        6.0   
4      26093.0   465.0       -2471.0     -3064.0       -70.0        2.0   
...        ...     ...           ...         ...         ...        ...   
54202  31216.0   515.0        -461.0      3035.0        45.0        5.0   
54203  28830.0   473.0        1728.0      -310.0         7.0       10.0   
54204  29140.0   466.0       -1728.0       310.0        -7.0        6.0   
54205  28479.0   492.0       -2693.0     -2278.0       -39.0        4.0   
54206  30757.0   531.0        2693.0      2278.0        39.0        7.0   

       assistsat15  
0              0.0  
1              2.0  
2              0.0  
3              7.0  
4              3.0  
...            ...  
54202          4.0  
54203         19.0  
54204         14.0  
54205          9.0  
54206         16.0  

[54207 rows x 51 columns]

--- 
Now we are going to creat 2 data frames based on the 2 questions we have; 

1. First data frame will look excatly as is and if it's needed we will adjust it accordingly on our second notebook


2. Second model will be only based on gold and experience or any relative features that gives us information regarding these 2.

Now lets create our cleaned dataframes and save them into our local machine so we can import them into our second model.

---

In [ ]:
# Importing dataframe into local files
M1_df.to_csv('folder/subfolder/out.csv')  

In [ ]:
# Creating our second data frame for golds and EXP
log_df = M1_df.drop(['dragons','side','firstdragon','dragons','infernals','mountains','clouds','oceans','chemtechs','hextechs','dragons (type unknown)','elders','firstherald','heralds','firstbaron','barons','firsttower','towers','firstmidtower','firsttothreetowers','turretplates','inhibitors','controlwardsbought'], axis=1)

In [ ]:
# Importing our second dataframe into local files
log_df.to_csv('folder/subfolder/out.csv')  

In [ ]:
# Sanity check
log_df

teamname  result  kills  assists  teamkills  firstblood  \
0       Liiv SANDBOX       1     17       33         17         0.0   
1                 T1       0     11       25         11         1.0   
2                 T1       0      5       16          5         0.0   
3       Liiv SANDBOX       1     21       41         21         1.0   
4                 T1       0      4        8          4         0.0   
...              ...     ...    ...      ...        ...         ...   
54202    QLASH Spain       1     22       26         22         1.0   
54203   Kawaii Kiwis       1     28       70         28         1.0   
54204  Stormbringers       0     10       24         10         0.0   
54205           ZEST       1     23       70         23         0.0   
54206    QLASH Spain       0     13       35         13         1.0   

       totalgold  earnedgold  earned gpm  goldspent  ...  killsat10  \
0          68927     44905.0   1213.6486    62275.0  ...        0.0   
1          62709     38687.0   1045.5946    62025.0  ...        0.0   
2          57283     33190.0    894.2075    56575.0  ...        0.0   
3          72004     47911.0   1290.8217    63975.0  ...        3.0   
4          50749     28981.0    869.8649    48400.0  ...        0.0   
...          ...         ...         ...        ...  ...        ...   
54202      65922     43073.0   1227.7340    58600.0  ...        3.0   
54203      67565     45389.0   1335.6253    62650.0  ...        7.0   
54204      56361     34185.0   1005.9343    53975.0  ...        2.0   
54205      68779     46001.0   1315.5672    63905.0  ...        2.0   
54206      58072     35294.0   1009.3613    54725.0  ...        3.0   

       assistsat10  goldat15   xpat15  csat15  golddiffat15  xpdiffat15  \
0              0.0   21953.0  27219.0   507.0        -825.0     -1665.0   
1              0.0   22778.0  28884.0   519.0         825.0      1665.0   
2              0.0   21714.0  26869.0   501.0       -5484.0     -4833.0   
3              3.0   27198.0  31702.0   557.0        5484.0      4833.0   
4              0.0   22329.0  26093.0   465.0       -2471.0     -3064.0   
...            ...       ...      ...     ...           ...         ...   
54202          2.0   24907.0  31216.0   515.0        -461.0      3035.0   
54203         12.0   26400.0  28830.0   473.0        1728.0      -310.0   
54204          3.0   24672.0  29140.0   466.0       -1728.0       310.0   
54205          4.0   23741.0  28479.0   492.0       -2693.0     -2278.0   
54206          6.0   26434.0  30757.0   531.0        2693.0      2278.0   

       csdiffat15  killsat15  assistsat15  
0           -12.0        0.0          0.0  
1            12.0        1.0          2.0  
2           -56.0        0.0          0.0  
3            56.0        6.0          7.0  
4           -70.0        2.0          3.0  
...           ...        ...          ...  
54202        45.0        5.0          4.0  
54203         7.0       10.0         19.0  
54204        -7.0        6.0         14.0  
54205       -39.0        4.0          9.0  
54206        39.0        7.0         16.0  

[54207 rows x 29 columns]

--- 
# Conclusion 


Now that we save our two cleaned dataframe and done with our EDA, we can go ahead and start new notebook and import our cleaned datasets so we can proceed to start building our models and analyse them.